src/main/java/com/palantir/refreshable/DefaultRefreshable.java

```diff
@@ -100,64 +111,81 @@ public T current() {
+            Disposable unsubscribeAndUntrack = () -> {
+                disposable.dispose();
+                rootSubscriberTracker.deleteReferenceTo(trackedSubscriber);
+            };
+            return unsubscribeAndUntrack;
+        } finally {
+            readLock.unlock();
+        }
     }
 
-    private synchronized Disposable subscribeToSelf(Consumer<? super T> subscriber) {
-        orderedSubscribers.add(subscriber);
-        int subscribers = orderedSubscribers.size();
-        if (subscribers > WARN_THRESHOLD) {
-            log.warn(
-                    "Refreshable {} has an excessive number of subscribers: {} and is likely leaking memory. "
-                            + "The current warning threshold is {}.",
-                    SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
-                    SafeArg.of("numSubscribers", subscribers),
-                    SafeArg.of("warningThreshold", WARN_THRESHOLD),
-                    new SafeRuntimeException("location"));
-        } else if (log.isDebugEnabled()) {
-            log.debug(
-                    "Added a subscription to refreshable {} resulting in {} subscriptions",
-                    SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
-                    SafeArg.of("numSubscribers", subscribers));
+    private void preSubscribeLogging() {
+        if (log.isWarnEnabled()) {
+            int subscribers = orderedSubscribers.size() + 1;
+            if (subscribers > WARN_THRESHOLD) {
+                log.warn(
+                        "Refreshable {} has an excessive number of subscribers: {} and is likely leaking memory. "
+                                + "The current warning threshold is {}.",
+                        SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
+                        SafeArg.of("numSubscribers", subscribers),
+                        SafeArg.of("warningThreshold", WARN_THRESHOLD),
+                        new SafeRuntimeException("location"));
+            } else if (log.isDebugEnabled()) {
+                log.debug(
+                        "Added a subscription to refreshable {} resulting in {} subscriptions",
+                        SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
+                        SafeArg.of("numSubscribers", subscribers));
+            }
         }
+    }
 
+    @GuardedBy("readLock")
+    private Disposable subscribeToSelf(Consumer<? super T> subscriber) {
+        preSubscribeLogging();
+        orderedSubscribers.add(subscriber);
         subscriber.accept(current);
-        return () -> remove(subscriber);
+        return new DefaultDisposable(orderedSubscribers, subscriber);
     }
 
-    private synchronized void remove(Consumer<? super T> subscriber) {
-        orderedSubscribers.remove(subscriber);
-    }
+    // DefaultDisposable ensures the resulting disposable doesn't accidentally reference a Subscriber object.
```

Comment:
hmm, too many words and I used the wrong ones!

A Disposable object shouldn't prevent a Refreshable from being garbage collected if the refreshable is no longer referenced -- in that case it's impossible that the consumer could be called with an update because the root refreshable has been garbage collected.

Reply:
Updated docs + example.